### ISSUE TITLE
DryIoc cold start performance

### ISSUE THREAD

**Issue by Fruchuxs** [dadhi/DryIoc#27]
After one week of investigations we found out, that DryIoc 3.0.2 is rather slow in comparison to 2.12.8 at the first execution time (cold start). In most cases the first call of a C# application doesn't count. But we measured a time increase of ca. six seconds for greater object graphs in our application. So for example our asp.net core 2.1 application starts, DryIoc resolves some small Services, than the applications tries to resolve a controller and .. we can make a cup of coffee. For a web applications this is just uncomfortable (because after the JIT compiling it's fast), but for one of our console applications this is really painful.

We looked at the expression trees which dryioc generates in version 3.0.2 and 2.12.8 for one of our object graphs. Too much to post here and to analyze in depth, but we can say, that the dryioc 3.0.2 expression tree is roundabout 90% bigger and seems even more complex.

We didn't changed the service registrations much. We just replaced Reuse.InWebrequest through Reuse.Scoped and two RegisterDelegates through RegisterMany as you mentioned out [here](https://bitbucket.org/dadhi/dryioc/issues/595/documentation-upgrade-from-dryioc2-to-3#comment-47209687). 

Maybe you can say more about this? In most of the cases we use simple service registrations like `Register<IA, A>`, or factories, or `made: Parameters.Of.Type` ... We also use the `IEnumerable`-resolve Feature.

**Comment by dadhi** [comment]
Hi @Fruchuxs ,

v3 does not reuse expressions across the object graph, so if you have `A` in multiple places in the object graph, you will have new expressions each time. Basically, the expression cache was removed, because it was really hard to figure out the invalidation pattern.

In upcoming v3.1 the expression reuse (cache) is back. But it has a different transient nature, which is simple to manage. It should improve cold-start performance.

For the other use-cases, I am not sure. Could you provide some notable parts of the object graphs, where you see an increase comparing to v2?

**Comment by dadhi** [comment]
You can use `container.Resolve<LambdaExpression>(serviceType)` to get an actual expression. Just don't forget to `OpenScope` if you are resolving a scoped service.

**Comment by Fruchuxs** [comment]
Hello again, thanks for your quick responses. I think I will await v3.1. With your explanations it makes sense, why the graphs seems more complex, because of the nested graphs. So no problems here I think. Can you give a time span you think you are ready with v3.1? Just to examinate if we need to downgrade.

**Comment by dadhi** [comment]
Next week, hopefully.

**Comment by dadhi** [comment]
Another idea, I can release a preview of 3.1 now. 

Would you be able to check performance using it?

**Comment by dadhi** [comment]
The late update..

While wirking in 3.1 preview release, I have found (remembered) other problem with .net core targeting. That .net core won't use FEC, oops :(

That was done close to v3 because of complexity (new projects needed) to support .netstandard 1.3 and 2.0 versions.

But the main obstacle is that I have used old csproj system.

I went ahead, and started to bring everything under new csprojs with multi-targeting. At the moment I am almost complete to release the preview. Hopefully, tomorrow.

**Comment by dadhi** [comment]
Preview is out: https://www.nuget.org/packages/DryIoc.dll/3.1.0-preview-01

**Comment by Fruchuxs** [comment]
Thank you for the prerelease.
But we got an exception in our main project we can't explain. I assume a problem with [
DryIoc.Microsoft.DependencyInjection](https://www.nuget.org/packages/DryIoc.Microsoft.DependencyInjection/2.1.0). A Testproject with only DryIoc works fine.

> System.IO.FileLoadException: 'Could not load file or assembly 'DryIoc, Version=3.0.2.0, Culture=neutral, PublicKeyToken=null'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)'

I checked all our projects and can't find a reference to 3.0.2.0, but VS generates 3.0.2.1 references in *.assembly.cache and *.deps.json files. Looks strange to me.

**Comment by dadhi** [comment]
Try the new DI.MS.DI preview:
https://www.nuget.org/packages/DryIoc.Microsoft.DependencyInjection/2.2.0-preview-01

**Comment by Fruchuxs** [comment]
Got the same exception with different assembly version number (3.1.0.0). We looked into the .net standard 2.0 assembly with ILSpy and it seems, that the assembly version is not set ("0.0.0.0"). Dunno how this happened, your cproj Files seems to be correct.

**Comment by dadhi** [comment]
But now I know a possible reason. Thank you. I will release new DryIoc preview 2 with fix.

**Comment by dadhi** [comment]
https://www.nuget.org/packages/DryIoc.dll/3.1.0-preview-02

**Comment by Fruchuxs** [comment]
Not so fast as 2.12.8, but a greatly increse to 3.0 in our cases (~0,5 seconds difference). Over all it looks great. Thank you.

**Comment by dadhi** [comment]
Interesting, 
I would like to look at your object graph. Maybe it something simple to fix or optimize.

**Comment by Fruchuxs** [comment]
Here it is.
[tree.txt](https://github.com/dadhi/DryIoc/files/2449264/tree.txt)

**Comment by dadhi** [comment]
Thank you very much. It is a deep tree indeed.

Could you try to add the rule to the container:
```cs
new Container(rules => rules.WithoutDependencyDepthToSplitObjectGraph());
```
or via:
```cs
container.With(rules => rules.WithoutDependencyDepthToSplitObjectGraph());
```
and check the results.

The reason is the default `Rules.DefaultDependencyDepthToSplitObjectGraph` is set to `8`. Maybe, it is not a good default value. It would be good to find out.

The number is used to split an object sub-graph deeper than the 8th level to the separate `Resolve` call, 
But this nested call requires a context from the upper graph (for instance, if you need a condition based injection, or find a mismatching lifestyle), so there goes call-stack recreation with `Request.Empty.Push(..).Push(..).Push(..),..`, etc. So, here is the tradeoff.

**Comment by Fruchuxs** [comment]
Thank you for the information and explanation. But with `WithoutDependencyDepthToSplitObjectGraph` the container resolves slower and the graph seems bigger (a little bit). In conclusion it doesn't matter.

**Comment by Fruchuxs** [comment]
Okay, I tested with another graph (but with a similar depth).
Here the results:

**WithDependencyDepthToSplitObjectGraph(20)**
00:00:00.1715299

**WithDependencyDepthToSplitObjectGraph(2)** // just for the science
00:00:00.1683051

**WithDependencyDepthToSplitObjectGraph(8)**
00:00:00.1934737
 
**With Default Constructor**
00:00:00.1531935
 
**WithoutDependencyDepthToSplitObjectGraph**
00:00:00.1361311

**Comment by Fruchuxs** [comment]
But the measurements fluctating .. As I said, I don't think it matters in our cases.

**Comment by dadhi** [comment]
Thank you for measuring!

Seems like using the depth split does not add any value in your case.

Need to go deeper..

**Comment by dadhi** [comment]
Could you use a ExpressionToCodeLib nuget package and dump the expression without depth argument?

**Comment by Fruchuxs** [comment]
Here the Tree To Code output from the tree I used for measuring.

[treetocode.txt](https://github.com/dadhi/DryIoc/files/2449902/treetocode.txt)

**Comment by dadhi** [comment]
That's what I need! Thank you.

**Comment by dadhi** [comment]
Shame on me, the FastExpressionCompiler was not included yet because of different compile constant :-(
Now fixed in DryIoc v3.1.0-preview-03.
Please, check.

**Comment by Fruchuxs** [comment]
Btw. (I don't tested pv03 yet) we investigated with pv02, that we got an lifespan exception in our web project:

> ContainerException: Dependency AuthStateManager as parameter "aAuthStateManager" reuse Scoped {Lifespan=100} lifespan shorter than its parent's: scoped AuthenticationEvents #284 from Container with Scope {no name} with Rules with {TrackingDisposableTransients, SelectLastRegisteredFactory} and without {ThrowOnRegisteringDisposableTransient} with FactorySelector=<non-default> with Made.Of=<non-default> To turn Off this error, specify the rule with new Container(rules => rules.WithoutThrowIfDependencyHasShorterReuseLifespan()).

The problem is, that AuthStateManager is added via dryIoc as reuse: Reuse.Scoped and AuthenticationEvents  via .net DI with AddScoped. We replaced the .net DI registration with a dryIoc equivalent and we don't got the exception anymore. We checked the Code in Di.Ms.Di and discovered that AddScoped is registred in dryIoc as Reuse.ScopedOrSingleton. So we changed our replaced service registration with Reuse.Scoped to Reuse.ScopedOrSingletone and the exception occured again. We tested out if our services was added as singleton by dryioc, because maybe no scope was open and the new version is just more restrictive, but we had to recognize that the object is created at every request and not just first.
In conclusion it seems, that dryioc means that ScopedOrSingleton is more healthy as Scoped which seems strange.

**Comment by Fruchuxs** [comment]
Don't know why, but prev02 seems faster.
I started a loop with 50 iterations. with the creation of the container and measured the resolving of the object from the graph I posted before ... The median as well as the average says, that prev02 is faster for this graph.


prev02 | prev03|type
------------ | -------------|------------ 
0,0178340 | 0,0214996|median
0,0224712 | 0,0261303|average

Maybe my test is nonsense, because I have no idea how your cache works. I just hope you cache is an instance no-shared cache and no static shared cache.
I think this values are negligible. Sadly I can't test the performance of the more complex tree of our web controllers good because of asp.net core dependencies and contexts.

**Comment by dadhi** [comment]
ok, thanks for that. And I am lost a bit. Let it sink in over weekend :)

**Comment by Fruchuxs** [comment]
Okay, we found out that there is some JIT magic .. If we resolve an object, the dryioc tree doesn't seems to be compiled by the JIT compiler, because JIT means, that it doesn't need to instatiate the object because it isn't used yet. First we access a method on the object, the JIT compiler seems to instantiate the object completly (but just with the dependencies he needs to execute the method). First then he starts to compile the dryioc tree in the need to instantiate the object ... I'm not sure this is true. But this means, that my tests are just nonesense. I will try to find a better test.

**Comment by dadhi** [comment]
Hi again,

Just realeased a Preview 4: https://www.nuget.org/packages/DryIoc.dll/3.1.0-preview-04

I have fixed (hopefully) bug with wrong reuse mismatch for Scoped and ScopedOrSingleton.

I have also condensed a Request stack passed to Resove to a single Constant, it should decrease size of graph with default depth split value (8).

Will be happy to hear your feeadback. Thanks!

**Comment by Fruchuxs** [comment]
Hey,

New performance numbers for you:

3.0 | pv2 | pv4
-- | -- | --
5,0888453 | 1,5168796 | 2,1876228
5,5643245 | 1,8527184 | 0,9721876
5,9362024 | 1,9551376 | 1,1744556
7,013724 | 2,0682004 | 1,3053725
7,4730085 | 2,1848985 | 2,2390254
  |   |  
6,21522094 | 1,9155669 | 1,57573278

We tested it with the whole console application which uses the dependency graph above, but the app doesn't do much. So the performance lost through other code should be marginal.
The other bug doesn't occur anymore since prev 04. :)

**Comment by dadhi** [comment]
@Fruchuxs Thanks, What about V2 for comparison?

**Comment by Fruchuxs** [comment]
Here the results incuding DryIoc 2.12.


3.0 | pv2 | pv4 | 2.12
-- | -- | -- | --
5,0888453 | 1,5168796 | 2,1876228 | 1,6066698
5,5643245 | 1,8527184 | 0,9721876 | 1,4519457
5,9362024 | 1,9551376 | 1,1744556 | 1,4118938
7,013724 | 2,0682004 | 1,3053725 | 1,4485652
7,4730085 | 2,1848985 | 2,2390254 | 1,2885061
  |   |   |  
6,21522094 | 1,9155669 | 1,57573278 | 1,44151612

It's hard to measure, there are hight flucatuations between the probes. I assume this is a problem with the CPU boost / throttling.

**Comment by dadhi** [comment]
Ok, now I can see the scale. 

I think, the major problem is addresed :)

So, I am planning to close the issue when v3.1 rtm is released.

The expression dump you provided gave me a ton of info, and hopefully I will find some time later to dig it more thoroughly.

If you have an idea, what other optimizations may be done, let's discuss.

**Comment by Fruchuxs** [comment]
Nice to hear! If I find other Problems or suggestion I will open another ticket like [here](https://github.com/dadhi/DryIoc/issues/26). I'm looking forward for 3.1 Release. :)

**Comment by ahydrax** [comment]
Env: `.NET 4.6.2; DryIoc.dll 3.0.2; selfhost owin/IIS Express`

Faced the same problem: I have a complex object graph (with deep nesting services and decorators). Cold start for some on my controllers takes about 1 minute (!) and memory footprint raises up to 13Gb. Profiling showed that 99% memory allocations and CPU are consumed by JIT.

I'm also getting `StackOverflowException` when running my app over IIS Express but stack trace shows that there are only ~30-40 stack frames. The most interesting thing is that my app works in selfhost mode. It seems like something inside DryIoC/JIT allocates a lot on stack and when I'm using IIS Express I'm hitting the stack size.

Gonna test 3.1.0-preview.

**Comment by dadhi** [comment]
Strange, not so much stack to choke in. 

Mmm, what is the size of graph? How many layers deep and how many arguments in constructors on each layer. What are the reuses, Scoped?

**Comment by ahydrax** [comment]
@dadhi,
no, I don't use any `Func<X>` or `Action<X>` dependencies. The most big controller (whom resolution causes `StackOverflowException`) uses 14 dependencies. Each dependency can have from 0 to 7 layers deep, from 1 to 15 constructor arguments, approximately ~200 resolutions have to be done in total. Most registrations are `Transient`. Controllers registered as `Scoped`.

**Comment by ahydrax** [comment]
@dadhi,
Found interesting thing, tried with both `3.0.2` and `3.1.0-preview5`
* Created container with default config
* Generated resolution expression using `container.GenerateResolutionExpressions(x => x.ServiceType.Name.EndsWith("MyControllerWithLotOfDependencies"));`
* Generated code using ExpressionToCodeLib and saved it to a file
`3.0.2` resulted a 5mb text file whereas `3.1.0-preview5` resulted 18mb file.

Anything else I can do to help you with diagnosting this?

**Comment by dadhi** [comment]
> 3.0.2 resulted a 5mb text file whereas 3.1.0-preview5 resulted 18mb file.

Omg, I suppose the latter result using `.WithoutDependencyDepthToSplitObjectGraph()` or not?

**Comment by dadhi** [comment]
Linking #39 and #40 here.

**Comment by dadhi** [comment]
What is the default output size in v3.1 without `.WithoutDependencyDepthToSplitObjectGraph()`

**Comment by ahydrax** [comment]
Same, I even tried to diff both files: exactly the same. ~17346kB

**Comment by dadhi** [comment]
Ok, could you try a different value for `.WithDependencyDepthToSplitObjectGraph(5)` like `5` or something.

**Comment by ahydrax** [comment]
Tried with `2` and `5`. Nothing changed. Could that be a problem with `ExpressionToCode` itself?

**Comment by dadhi** [comment]
Maybe... try to do just `Expression.ToString()`

**Comment by ahydrax** [comment]
Tried with `ToString()`. Got 21mb file for `2`, `5`, `8`, `Without`.

**Comment by dadhi** [comment]
Ohh, sorry there is `container.GenerateResolutionExpressions(...)`  maybe not the best tool, cause it tries to remove the run-time optimizations for generational scenarios. Instead, try to use:
```cs
var expr = container.Resolve<LambdaExpression>(typeof(MyControllerWithLotOfDependencies));
```

You need to open a scope if `MyControllerWithLotOfDependencies` registered with scope reuse:

```cs
using (var scope = container.OpenScope(Reuse.WebRequestScopeName))
    var expr = scope.Resolve<LambdaExpression>(typeof(MyControllerWithLotOfDependencies));
```

**Comment by ahydrax** [comment]
@dadhi 
Finally got different results:
`without` - 14mb
`8` - 3.7mb
`5` - 937kb
`2` - 6kb

Does that mean that I should use `2`? How it's connected with other aspects: lifetime control/peformance/etc?

**Comment by dadhi** [comment]
Thanks for results. The value is greatly depends on context. Like how many and what are types of reuse, do you have a lazy/func deps, how deep or wide a graph. Try to peek some sensible value between 5 and 8 for instance. Previous DryIoc default was 8 (v3.0). I did other changes to decrease a resilt expression size (see the whole thread). I am planning a new changes/options as well (#40). 

Maybe I will return back the default 8. So your and @Flo inputs are very valuable here.

**Comment by dadhi** [comment]
I wanted to add on resolving as `LambdaExpression`. You may experiment further by compiling the expression and (optionally) caching it in whatever suitable data structure. This way you are excluding all the container "magic"  and use the IoC container as an automated object creation tool.

```cs
var expr = scope.Resolve<LambdaExpression>(typeof(MyControllerWithLotOfDependencies));

var factory = (FactoryDelegate)expr.Compile(); 
var controller = (MyControllerWithLotOfDependencies)factory(container);
```
Btw, this also excludes FEC from the equatation, see #39.

**Comment by dadhi** [comment]
Just in case, I would also love to see a speed/memory comparison on your real graph between `expr.Compile()` and `expr.CompileFast()`.

**Comment by ahydrax** [comment]
Hi @dadhi,

As you requested, here is benchmark of `expr.Compile()` and `expr.CompileFast()`.
The benchmark code is [here](https://gist.github.com/ahydrax/d6e42f6fbb2f6c9bdb3ecf525016bc5a).

Benchmark parameters:

```BenchmarkDotNet=v0.11.2, OS=Windows 10.0.17763.55 (1809/October2018Update/Redstone5)
Intel Core i7-7700K CPU 4.20GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3190.0
  Job-MZJNWR : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3190.0

InvocationCount=10001  IterationCount=2  LaunchCount=1
RunStrategy=Monitoring  UnrollFactor=1  WarmupCount=1
```

With `.WithDependencyDepthToSplitObjectGraph(2)`
```
        Method |      Mean | Error |    StdDev | Ratio | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
-------------- |----------:|------:|----------:|------:|------------:|------------:|------------:|--------------------:|
 CompileNormal | 634.05 us |    NA | 1.1571 us |  1.00 |      6.5993 |      2.1998 |           - |            27.07 KB |
   CompileFast |  82.54 us |    NA | 0.0558 us |  0.13 |      6.6993 |      3.2997 |      0.2000 |            27.47 KB |
```

With `.WithDependencyDepthToSplitObjectGraph(3)`
```
        Method |     Mean | Error |   StdDev | Ratio | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
-------------- |---------:|------:|---------:|------:|------------:|------------:|------------:|--------------------:|
 CompileNormal | 985.9 us |    NA | 8.852 us |  1.00 |     30.3970 |           - |           - |           124.76 KB |
   CompileFast | 714.4 us |    NA | 7.205 us |  0.72 |     90.5909 |     46.6953 |      5.7994 |           359.43 KB |
```

With `.WithDependencyDepthToSplitObjectGraph(5)`
```
        Method |      Mean | Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
-------------- |----------:|------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
 CompileNormal |  4.312 ms |    NA | 0.0017 ms |  1.00 |    0.00 |    172.8272 |     85.9141 |           - |             1.04 MB |
   CompileFast | 17.192 ms |    NA | 0.0625 ms |  3.99 |    0.02 |   2928.0719 |    161.8382 |     52.9471 |             12.3 MB |
```

With `.WithDependencyDepthToSplitObjectGraph(7)`
```
        Method |     Mean | Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
-------------- |---------:|------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
 CompileNormal | 11.17 ms |    NA | 0.1530 ms |  1.00 |    0.00 |    504.9900 |    221.5569 |     75.8483 |             2.52 MB |
   CompileFast | 87.91 ms |    NA | 0.3260 ms |  7.87 |    0.08 |  16988.0240 |   1221.5569 |    189.6208 |            71.72 MB |
```

Seems that `CompileFast` grows exponentially depending on depth of graph.

**Comment by dadhi** [comment]
Thanks for the tests, very interesting. I would try the latest FEC v2 (after it is released) and will decide what to do.

Btw, seems that actual problems in your case was caused by FEC. Hopefully, will fix this soon(ish).

**Comment by dzmitry-lahoda** [comment]
May be related to https://github.com/dadhi/FastExpressionCompiler/issues/89#issuecomment-414571680.

Once I proposed to get real graphs into benchmark, created issue for voting, but it was closed without consideration https://github.com/danielpalme/IocPerformance/issues/86.

**Comment by dzmitry-lahoda** [comment]
Sot it next will not work for some reasons?
```
var expr = scope.Resolve<FastExpressionCompiler.LightExpression.LambdaExpression>(typeof(MyControllerWithLotOfDependencies)
```

**Comment by dadhi** [comment]
`FEC.LightExpression` is not supported cause FEC 2.0 is not yet integrared into DryIoc. Originally, the `ExpressionInfo` was not supported as a wrapper, as I did not think it was useful for the end client. After integration with FEC 2.0 maybe I will reconsider.. maybe

**Comment by dadhi** [comment]
@ahydrax,

I have released v3.1-preview-01 which sets a default depth to 5 for now, and moreover, simplifies the expressions with root scoped service (controller in your case). The result expression should have one less nested lambda (minus one compile step).

**Comment by ahydrax** [comment]
Hi @dadhi, here is benchmark results for `preview-06`.

```
BenchmarkDotNet=v0.11.2, OS=Windows 10.0.17763.55 (1809/October2018Update/Redstone5)
Intel Core i7-7700K CPU 4.20GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3190.0
  Job-GOPLAQ : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3190.0

InvocationCount=1001  IterationCount=2  LaunchCount=1
RunStrategy=Monitoring  UnrollFactor=1  WarmupCount=1

        Method |      Mean | Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
-------------- |----------:|------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
 CompileNormal |  4.138 ms |    NA | 0.0266 ms |  1.00 |    0.00 |    172.8272 |     85.9141 |           - |             1.04 MB |
   CompileFast | 16.984 ms |    NA | 0.0015 ms |  4.10 |    0.03 |   2578.4216 |    162.8372 |     52.9471 |            12.29 MB |
```

My observations (container created with `Rules.Default`):
* My app doesn't crash anymore with `StackOverflowException` under IIS Express
* Memory footprint drastically decreased (~1Gb instead of 4Gb and more - had the same with Autofac, I consider it's ok)

**Comment by Fruchuxs** [comment]
> Yes, using `.WithoutDependencyDepthToSplitObjectGraph()` on `3.1.0-preview` gives bigger output than stable version.

As I mentioed before, but our graph is not as big as @ahydrax Graph.

> uses 14 dependencies

I don't want to critisize your code nor your approaches, but with 14 dependencies there are maybe other problems here. If you actions uses all of these dependencies (high cohesion) it's okay, but if not, you shall split your controller into different ones. The Dependency injection pattern respectively the inversion of control principle assumes, that your classes fulfill the single responsibility principle.

This doesn't means, that DryIoc has problems here, but an IoC system has some constraints.

**Comment by ahydrax** [comment]
> I don't want to critisize your code nor your approaches, but with 14 dependencies there are maybe other problems here. If your actions uses all of these dependencies (high cohesion) it's okay, but if not, you shall split your controller into different ones. The Dependency injection pattern respectively the inversion of control principle assumes, that your classes fulfill the single responsibility principle.

I agree with you: having 14 dependencies might be an indicator that something wrong with the code itself.
Anyway, I didn't expect that changing DI container will break overall performance and/or its stability, especially that DryIoC is positioned as more performant container.

**Comment by dadhi** [comment]
Performance of containers depends on the use-case. Here, we likely hit the bug + specific case situation. But I glad for feetback, cause it pushed the boundaries for me.

**Comment by ahydrax** [comment]
Hi @dadhi , sure, approximately on Monday I'll post the results.

**Comment by CoskunSunali** [comment]
@ahydrax Do you have any updates to share with us?

**Comment by dadhi** [comment]
Here is the latest results: https://github.com/dadhi/DryIoc/issues/26#issuecomment-466460255

I will consider this issue closed, please open a new one if you see the problem.
